Display X-ABDATE, X-RELATEDNAME fields. Display labels for URLs

d7415 opened the pull request:
- Include associated X-ABLABELs with URLs.
- Display X-ABDATE and X-RELATEDNAME fields (with associated X-ABLABELs).

Tested against Google contacts. Probably not the neatest, and I haven't added anything to the template function yet.

scheibler commented:
In general the vcard standard allows private attributes. These must start with an X. At the moment
khard deals with them by adding them under the "private_objects" option in khard's config file
(khard.conf).

For example the line:

```
private_objects = Jabber, Twitter
```

Would process the vcard attributes "X-TWITTER" and "X-JABBER" and list them under "Private".

Please try to include yours like this:

```
private_objects = ABDATE, RELATEDNAME
```

The X-ABLABEL is something special. At the moment it's supported by phone, email and post address
cause it's very widely used there and difficult to ignore. But in general I want to minimize the
usage of these private attributes as much as possible.

Could you provide a sample vcard file that uses your custom attributes?

d7415 commented:
Thanks for responding.

I guess I didn't read enough before starting as I missed the `private_objects` config. That's certainly a neater way of handling them. Some method of aliasing might be nice though.

That just leaves the X-ABLABEL. I agree in principle with keeping private, unofficial attributes separate, but as the handling is already in place for those 3 areas, maybe making it more generic would be ok? Even if it was only for attributes in private_objects, and/or configurable.

I'll create a "clean" google contact, post it and export it when I get the chance (hopefully tonight)

d7415 commented:
Sample contact created in Google, retrieved by CARDDAV (vdirsyncer).
```
BEGIN:VCARD
VERSION:3.0
N:Smith;John;;;
FN:John Smith
REV:2019-01-15T23:02:05Z
UID:15fb7fec8d2ea347
EMAIL;TYPE=WORK:john.smith@example.com
EMAIL;TYPE=HOME:jsmith12345@example.net
item1.X-ABDATE:2019-01-15
item1.X-ABLabel:First Met
item2.X-ABDATE:2017-03-24
item2.X-ABLabel:Dog's Birthday
item3.X-ABRELATEDNAMES:Jane
item3.X-ABLabel:Wife
item4.X-ABRELATEDNAMES:Jimmy
item4.X-ABLabel:Child
item5.URL:https\://example.com
item5.X-ABLabel:Work
item6.URL:https\://example.net/john.smith
item6.X-ABLabel:Blog
END:VCARD
```

d7415 commented:
Current output (with `private_objects = Jabber, Skype, Twitter, ABDATE, ABRELATEDNAMES`):
```
Name: John Smith
Address book: google
E-Mail
    HOME: jsmith12345@example.net
    WORK: john.smith@example.com
Private:
    ABDATE: 
        - 2017-03-24
        - 2019-01-15
    ABRELATEDNAMES: 
        - Jane
        - Jimmy
Miscellaneous
    UID: 15fb7fec8d2ea347
    Webpage: 
        - https\://example.com
        - https\://example.net/john.smith
```

Output after (revised) PR:
```
Name: John Smith
Address book: google
E-Mail
    HOME: jsmith12345@example.net
    WORK: john.smith@example.com
Private:
    ABDATE: 
        - Dog's Birthday: 2017-03-24
        - First Met: 2019-01-15
    ABRELATEDNAMES: 
        - Child: Jimmy
        - Wife: Jane
Miscellaneous
    UID: 15fb7fec8d2ea347
    Webpage: 
        - Blog: https\://example.net/john.smith
        - Work: https\://example.com
```
It loses the localisation of ABDATE from the original PR, but that's not a big issue for me at least.

lucc commented:
As far as I can see you only show them in the pretty printed output. Do you plan to add them to the yaml output to make them editable? Is the current format parseable and unique when written to yaml? I suspect not. I would be puzzeled as a new user if I found data in the pretty printed output that I can not edit in the yaml.

More generally I think any line can be part of a group. Maybe we can even think about a more generic way to show them in the yaml format. Or we decide to not show them at all and just handle some very special and well defined cases of groups.

d7415 commented:
Sorry, I need to be clearer. I intend to add it everywhere, but didn't want to go too far down the rabbit hole if it's not something you're interested in (or the wrong way of doing it).

From what I've read it seems anything can be part of a group like you say. Generic handling would probably be useful, but could again be confusing if it moves fields (e.g. addresses) away from where they "usually" are.

Suggestion:

- Add `_get_group_count` and `_get_label` methods
- For any attribute, check for group name and count. If it's >2 treat it as a group, =2 check for a label and display it as a label (if present). If no label, treat as a group. If it has no group or group_count =1, ignore the group.

I'll still have to look at how adding/editing would be best done.

Does this sound sensible?

lucc commented:
I think your suggestions are mostly good. The result I would aim for at them moment is only a partial implementation of group handling for now. But we could also implement some building blocks for a more general handling. If we come up with use cases later we already have that.

This is how I would adjust your plan:
* Treating groups of sizes the same as no group is reasonable.
* for any groupsize > 2 we could just print a logging message of the debug level for now and then treat the group as non existent
* for group size = 2 we could do try to implement a special handling of `X-ABLABEL`. If there is no `X-ABLABEL` in that group we can again fall back to a simple log message and ignore the group.

There are obviously some corner cases:
1. groupsize 2, both are `X-ABLABEL` -> ignore group.
2. do we implement special treatment of `X-ABLABEL` for all fields or only for some? How would we represent the label on a `N` or `ADR` field in the yaml output (which must be parsable)?
3. What if a user has `ABLABEL` in the `private_objects` setting?

Final question: Where would you want to add the `_get_group_count` and `_get_label` methods? With which signatures?

d7415 commented:
Sorry. I haven't had much time to look at this.

> This is how I would adjust your plan:

That all sounds sensible.

>     2. do we implement special treatment of `X-ABLABEL` for all fields or only for some? How would we represent the label on a `N` or `ADR` field in the yaml output (which must be parsable)?

I would think as much as possible, but there might be areas where this would be unreasonable. As it will probably have to be added per field type, I'll look at these as I go. I'll have to read up on yaml too to check parsability and more into VCARD to check whether anything can't be grouped.

>     3. What if a user has `ABLABEL` in the `private_objects` setting?

Personally I'd look to put any ungrouped ABLABELs here, maybe with a warning. Or we could specifically exclude them with a warning?

> Final question: Where would you want to add the `_get_group_count` and `_get_label` methods? With which signatures?

Thinking about this again it might be easier (if it will be repeated a lot) to simply have a `_get_label(group.value)` method, returning `"label: "` or `""` as required (and outputting the debug information.
As for where, possibly up near `_get_string_field`. I was aiming for the `CarddavObject` scope to save passing that around when it's called.

lucc commented:
I think `_get_group_count` makes sense as a method on the `CarddavObject` class. But the thing that I do not understand is `_get_label`: Is it a function or a method? If it is a method, on which class? Which parameters does it have,  what are their types? What is the return type? What kind of side effects does it have (you clarified this one: debug output). 

At least that is the kind of information I would expect from a signature. I am only asking because for something that is called `_get_label` I feel that I can not deduce that myself. For the thing you named `_get_group_count` I would have deduced myself, that it is a method on `CarddavObject` without any parameters, that returns an `int`, namely the number of groups on that carddav object. Sorry for being unclear before.

d7415 commented:
Ah, I see what you mean. No problem - I was also unclear. `_get_ablabel` might be better. I'll try putting it into the format used in the source:

```py
    def _get_ablabel(self, group):
        """Get an ABLABEL for a specified group in the vCard.
        Will return the ABLABEL only if the group contains exactly two items, 
        exactly one of which is an ABLABEL.

        :param group: the name of the group
        :type group: str
        :returns: the ABLABEL in the circumstances above or an empty string
        :rtype: str

        """
```

d7415 commented:
Ok. This might need some looking at, but I have tested:
 - Displaying ABLABELs for URL, X-RELATEDNAMES, X-ABDATE
 - Editing the ABLABEL for a URL from `khard ed John Smith`
 - Adding a new X-RELATEDNAMES field with a label from `khard ed John Smith`

d7415 commented:
Any idea when you might have time to review/merge this?

No rush - I appreciate this work is unpaid! Just curious.

lucc commented on khard/carddav_object.py:
```diff
@@ -656,31 +703,57 @@ def _get_private_objects(self):
         webpage_obj = self.vcard.add('url')
-        webpage_obj.value = helpers.convert_to_vcard("webpage", webpage,
-                                                     ObjectType.string)
+        if isinstance(webpage, dict):
```
This looks like the same code as above. Does it make sense to add a wrapper function, something like `convert_to_vcard_with_ablabel()`?

d7415 replied:
Added as `_add_labelled_object` in #208.

lucc commented:
Also what do you think: Can you turn the contact you pasted above for demonstration into some test for the new code or the changed behaviour?

lucc commented:
I took the liberty to change the target to point to `develop` as that is where new development happens.

d7415 commented on khard/carddav_object.py:
```diff
@@ -656,31 +703,57 @@ def _get_private_objects(self):
+        if isinstance(value, dict):
+            if len(value) > 1:
+                raise ValueError(
+                    "Error: " + key + " must be a string.")
```
Good question. That makes no sense. I'll have to figure out what I was thinking here...

d7415 replied:
I think I was struggling to find wording that reflected what the user sees. I settled on
```py3
                raise ValueError("Error: %s must be a string or a dict " +\
                                 "containing one key/value pair." % obj_type)
```

d7415 commented on khard/carddav_object.py:
```diff
@@ -656,31 +703,57 @@ def _get_private_objects(self):
+            if len(value) > 1:
+                raise ValueError(
+                    "Error: " + key + " must be a string.")
+            label = [i for i in value][0]
```
I think I tried that, but it can't be indexed.

```py
TypeError: 'dict_keys' object does not support indexing
```

d7415 commented:
> I took the liberty to change the target to point to develop as that is where new development happens.

I noticed that, but not until I'd make the last set of changes, so I thought I'd wait and see. I've done most of the deconflicting locally now. Just going to finish that off and try to address other comments and then I'll push.

> Also what do you think: Can you turn the contact you pasted above for demonstration into some test for the new code or the changed behaviour?

I'll take a look at this. I'm not great at tests, but I do want to learn more.

d7415 commented:
Some days I hate git. But github more. I'm going to have to recreate this PR.

lucc commented:
Was the trouble introduced by me switching the target of this PR? I am sorry about that, maybe I can help you with some git commands if you describe the problem?

d7415 commented:
> Was the trouble introduced by me switching the target of this PR? I am sorry about that, maybe I can help you with some git commands if you describe the problem?

No. I merged the changes in, but was struggling to merge them back onto develop cleanly. Well, actually, to merge the develop commits into my branch, that onto master (because that's where this PR is from and can't be changed) and *then* back onto develop.

I opened #208 and it's fine.

d7415 commented:
> > Was the trouble introduced by me switching the target of this PR? I am sorry about that, maybe I can help you with some git commands if you describe the problem?
> 
> No. I merged the changes in, but was struggling to merge them back onto develop cleanly. Well, actually, to merge the develop commits into my branch, that onto master (because that's where this PR is from and can't be changed) and _then_ back onto develop.
> 
> I opened #208 and it's fine.

(the first commit there is merging this with develop, possible accidentally fixing things. The others are addressing your comments.)

lucc commented:
In such cases `git rebase` is often helpful. And you can also `git push -f` to a PR (*never ever `push -f` to a "published" branch like `master` or `develop` or so!!*).  Just in case you did not know these commands.

I will review the new PRs tomorrow.

d7415 commented:
Yeah, I force-pushed earlier in this PR :) I think my mistake was in the order of things. Git wasn't recognising my merge commit properly. I might have been able to get around it, but I didn't really want a long-lived PR on my local master anyway - that'll teach me :P
